### ISSUE TITLE
znode freed while still in use

### ISSUE THREAD

**Issue by lundman** [openzfsonwindows/ZFSin#89]
Crash after CrystalDiskMark

**Comment by jheuking** [comment]
hm, would be nice to have had some printf when znodes get freed...
can you reproduce that error?

**Comment by jheuking** [comment]
ah got that error too, just after rebasing on master

**Comment by jheuking** [comment]
so, as it seems, the znode gets freed through IRP_MJ_CLEANUP->delete_entry->....blablabla->zfs_remove

and then goes and deletes the znode here everytime : https://github.com/openzfsonwindows/ZFSin/blob/5f001c8851ef264e162c762b5d275bddb69fa7c3/ZFSin/zfs/module/zfs/zfs_vnops.c#L2154

Guess this is a little bit problematic, because we can't release the znode before the vnode in any case if I got that right? So if we don't recycle the vnode there, like it was before, we shouldn't delete the znode either.
So without knowing a better solution, I would say that part always leads to a crash and should definitely be changed.
 
What I observed is that in my case usually vp->v_iocount = 2 and vp->v_usecount = 0. That wouldn't be too bad, if vnode_isinuse would consider v_iocount as well as v_usecount.
If we would get vnode_isinuse == true, delete_now wouldn't get set and we would just get to the unlinked_list.

In summary I would say we only have two options. We can either remove the vnode directly in zfs_remove like it was before and free the znode afterwards. Or if we don't want to free the vnode there, we can't free the znode there neither and have to put it on the unlinked list.

**Comment by lundman** [comment]
Because
A) Windows requires the file to be deleted at CLEANUP time (the fsdtester checks for it)
B) The vnode can't be released until CLOSE

So, we should in CLEANUP, if vnode_unlink() is set, "signal" that the file was deleted - probably by setting a flag in vnode.
Then in `zfs_vnop_lookup()` if we get the vnode, but it is signaled deleted, return ENOENT. 
Finally, in CLOSE we actually call `delete_entry` which can delete vp and zp as normal.

So, vnode needs a "delete on close" flag, and a "already deleted" flag. Then move `delete_entry` from CLEANUP, to CLOSE, and just "pretend" the file is deleted between the two events CLEANUP and CLOSE.

**Comment by lundman** [comment]
Ok, so reading http://www.osronline.com/article.cfm?article=102 - I first tried to implement the second "easier" approach, but clearly that is not working.

So we need to hybrid it with the AVL tree I think. If we see a `FileObject->Flags&FO_STREAM_FILE` and if not already present, needs to be inserted into the AVLtree and `usecount++` (to stop recycle).

Then when receiving IRP_CLOSE, check for AVLtree existence, if so remove it and `usecount--`.

I assume that FO_STREAM_FILE check should be done in `fs_dispatcher` for any valid IRP.

We should presumably store the AVLtree inside the vnode - unique for that vnode - of just the FO. Alternatively, one global tree storing [FO,vnode] pairs.

**Comment by jheuking** [comment]
Would prefer to store the tree in the vnode struct, guess that works faster...

**Comment by lundman** [comment]
Been fixed in the great-vnode-rewrite of '18.
